blueocean-dashboard/src/main/js/components/karaoke/components/FreeStyle.jsx

```diff
@@ -0,0 +1,89 @@
+import React, { Component, PropTypes } from 'react';
+import { logging } from '@jenkins-cd/blueocean-core-js';
+import { observer } from 'mobx-react';
+import { KaraokeService } from '../index';
+import LogConsole from './LogConsole';
+import LogToolbar from './LogToolbar';
+
+const logger = logging.logger('io.jenkins.blueocean.dashboard.karaoke.FreeStyle');
+
+@observer
+export default class FreeStyle extends Component {
+    /**
+     * Mainly implemented because
+     * - we need to fetch the log to display it
+     */
+    componentWillMount() {
+        if (this.props.augmenter) {
+            this.fetchData(this.props);
+        }
+    }
+    /**
+     * Mainly implemented because
+     * - we need to re-fetch the log to display it when the log finished
+     * - or when start is set to 0
+     */
+    componentWillReceiveProps(nextProps) {
+        const nextStart = nextProps.location && nextProps.location.query ? nextProps.location.query.start : undefined;
+        const currentStart = this.props.location && this.props.location.query ? this.props.location.query.start : undefined;
+        logger.debug('newProps mate', nextProps, currentStart);
+        if (!nextProps.augmenter.karaoke) {
+            this.stopKaraoke();
+        }
+        if (
+            (currentStart !== nextStart && nextStart !== undefined) ||
+            (nextProps.run.isCompleted() && !this.props.run.isCompleted())
+        ) {
+            logger.debug('re-fetching since result changed and we want to display the full log');
+            this.pager.fetchGeneralLog({ start: nextStart });
+        }
+    }
+    componentWillUnmount() {
+        this.stopKaraoke();
+    }
+    stopKaraoke() {
+        logger.debug('stopping karaoke mode, by removing the timeouts on the pager.');
+        this.pager.clear();
+    }
+    fetchData(props) {
+        const { augmenter, location } = props;
+        this.pager = KaraokeService.generalLogPager(augmenter, location);
+    }
+    render() {
+        if (this.pager.pending) {
+            logger.debug('abort due to pager pending');
+            return null;
+        }
+        const { t, router, location, scrollToBottom, augmenter } = this.props;
+        const { data: logArray, hasMore } = this.pager.log;
+        logger.debug('props', scrollToBottom, this.pager.log.newStart, augmenter.generalLogUrl);
```

Comment:
render should be artifact-free and could be called repeatedly, even if it's not logging things to console due to the level, it is still creating objects that just get destroyed for no real reason. better to log things when state changes, I think, than just logging what is happening. that's what debugging is for.